staging/src/k8s.io/client-go/discovery/cached/disk/cached_discovery.go

```diff
@@ -100,35 +105,77 @@ func (d *CachedDiscoveryClient) ServerGroupsAndResources() ([]*metav1.APIGroup,
 	return discovery.ServerGroupsAndResources(d)
 }
 
-// ServerGroups returns the supported groups, with information like supported versions and the
-// preferred version.
-func (d *CachedDiscoveryClient) ServerGroups() (*metav1.APIGroupList, error) {
-	filename := filepath.Join(d.cacheDirectory, "servergroups.json")
+func (d *CachedDiscoveryClient) GroupsAndMaybeResources() (
+	*metav1.APIGroupList, map[schema.GroupVersion]*metav1.APIResourceList, error) {
+	// First attempt to retrieve the cached aggregated discovery groups and resources file.
+	filename := filepath.Join(d.cacheDirectory, "servergroupsandresources.json")
 	cachedBytes, err := d.getCachedFile(filename)
+	if err == nil {
+		cachedGroupsAndResources := apidiscovery.APIGroupDiscoveryList{}
+		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), cachedBytes, &cachedGroupsAndResources); err == nil {
+			klog.V(10).Infof("returning cached discovery info from %v", filename)
+			// Parse the APIDiscoveryGroupList into groups and resources map by group/version.
+			cachedGroups, cachedResourcesMap := discovery.ParseAggregatedDiscovery(cachedGroupsAndResources)
+			return cachedGroups, cachedResourcesMap, nil
+		}
+	}
+
+	// No cached aggregated discovery, so attempt cached previous discovery format.
+	filename = filepath.Join(d.cacheDirectory, "servergroups.json")
```

Comment:
Only one filename now. No longer caching `servergroupsandresources.json` explicitly, since it will be cached in the `http` cache if server understands aggregated discovery.